docs/proposals/scheduling/20220609-resource-reservation.md

```diff
@@ -0,0 +1,233 @@
+
+![image](/docs/images/resource-reservation.svg)
+
+```go
+type Reservation struct {
+	metav1.TypeMeta `json:",inline"`
+	// <unresolved>
+	// Should the reservation be namespaced?
+	// </unresolved>
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ReservationSpec   `json:"spec,omitempty"`
+	Status            ReservationStatus `json:"status,omitempty"`
+}
+
+type ReservationSpec struct {
+	// Template defines the scheduling requirements (resources, affinities, images, ...) processed by the scheduler just like a normal pod.
+	// If the `template.spec.nodeName` is specified, the scheduler will not choose another node but reserve resources on the specified node.
+	Template *corev1.PodTemplateSpec `json:"template,omitempty"`
+	// Select the owners that can allocate the reserved resources.
+	// Multiple selectors and ANDed.
+	OwnerSelectors []ReservationOwnerSelector `json:"ownerSelectors,omitempty"`
+	// Time-to-Live period (in seconds) for the reservation.
+	// If both `TTLSeconds` and `ExpiredTimestamp` are not specified, a very long TTL will be picked by default.
+	TTLSeconds *int64 `json:"ttlSeconds,omitempty"`
+	// Expired time for the reservation.
+	// If specifed, the `TTLSeconds` will be ignored.
+	ExpiredTimestamp metav1.Time `json:"expiredTimestamp,omitempty"`
+	// If specified, allocate resources from the referenced objects.
+	ResourceFrom []ReservationOwner `json:"resourceFrom,omitempty"`
+	// By default, reservation's resources requirements (specified in `template.spec`) sums resources would get eventually (specified in `ResourceFrom`) should be not larger than the free resources of the node.
+	// Once `SkipResourceValidate` is set, this validation will be skipped and then overcommitment is allowed.
+	SkipResourceValidate bool `json:"skipResourceValidate,omitempty"`
+}
+
+type ReservationStatus struct {
+	// The `Phase` indicates whether is reservation is waiting for process (`Pending`), available to allocate (`Available`) or expired to get cleanup (Expired).
+	Phase ReservationPhase `json:"phase,omitempty"`
+	// The `Conditions` indicate the messages of reason why the reservation is still pending.
+	Conditions []ReservationCondition `json:"conditions,omitempty"`
+	// Current resource owners which allocated the reservation resources.
+	Owners []ReservationOwner `json:"owners,omitempty"`
+}
+
+type ReservationOwner struct {
+	metav1.OwnerReference
+	// <unresolved>
+	// The `Namespace` field is needed when Reservation is not namespaced.
+	// </unresolved>
+	Namespace string `json:"namespace,omitempty"`
+}
+
+type ReservationOwnerSelector struct {
+	// Multiple field selectors are ORed.
+	ReservationOwner
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+	// Reference a controlled object.
+	ControllerKey ControllerReference `json:"controllerKey,omitempty"`
```

Comment:
consider using OwnerReference ？

Reply:
Fixed. If `Reservation` is non-namespaced, we also have to extend `OwnerReference` with a field `Namespace`.